In [23]:
import pandas as pd
import numpy as np

name_file='Dataset_11_poke.csv'
err_name='Fichier inconnu'

# Treatment_1 prend en argument le nom du fichier csv à traiter, 
# Renvoie un dataFrame de ce fichier (df)
def Treatment_1(name=""):
    try:
        df=pd.read_csv(name_file,sep='|')
    except:
        print(err_name) 
	#Retourner le dataframe
    return df
df=Treatment_1(name_file)

In [2]:
df.shape

(801, 52)

In [3]:
df.columns

Index(['Unnamed: 0', 'Unnamed: 0.1', 'Unnamed: 0.1.1', 'Unnamed: 0.1.1.1',
       'Unnamed: 0.1.1.1.1', 'Unnamed: 0.1.1.1.1.1', 'Unnamed: 0.1.1.1.1.1.1',
       'Unnamed: 0.1.1.1.1.1.1.1', 'Unnamed: 0.1.1.1.1.1.1.1.1',
       'Unnamed: 0.1.1.1.1.1.1.1.1.1', 'Unnamed: 0.1.1.1.1.1.1.1.1.1.1',
       'abilities', 'against_bug', 'against_dark', 'against_dragon',
       'against_electric', 'against_fairy', 'against_fight', 'against_fire',
       'against_flying', 'against_ghost', 'against_grass', 'against_ground',
       'against_ice', 'against_normal', 'against_poison', 'against_psychic',
       'against_rock', 'against_steel', 'against_water', 'attack',
       'base_egg_steps', 'base_happiness', 'base_total', 'capture_rate',
       'classfication', 'defense', 'experience_growth', 'height_m', 'hp',
       'japanese_name', 'name', 'percentage_male', 'pokedex_number',
       'sp_attack', 'sp_defense', 'speed', 'type1', 'type2', 'weight_kg',
       'generation', 'is_legendary'],
      dtype='

Dans le cadre de ce projet, nous allons analyser et traiter une base de donnée contenant
des informations sur les Pokémons,
Le dataset 7 generations de pokémon avec 801 instances et 52 attributs
On va traiter principalement 

In [4]:
df['generation'].unique()

array([1, 2, 3, 4, 5, 6, 7])

Supprimer les doublons 

In [5]:
df['name'].unique().shape
#On remarque que le nombre de noms de pokémon differents est egale au nombre de lignes du dataset 
#Donc il n ya pas de doublons !

(801,)

In [6]:
# Après application de la méthode suivante, on aura le meme dataset (Aucune ligne ne sera supprimée)

In [7]:
df=df.drop_duplicates(subset=None, keep='first', inplace=False)
df.shape

(801, 52)

Numériser l'ensemble des colonnes de manière pertinentes 

In [8]:
#Declaration des fonctions 
def replace_row_nan(row):
    try:
        for i in df.columns:
            if i in row:
                for j in df.index:
                    try:
                        if float(df[i][j]) > df.shape[0] or float(df[i][j]) <0:
                            df[i][j]=np.nan
                    except:
                        df[i][j]=np.nan
    except:
        print("error")
        
    return df

In [9]:
#Traitement des colonnes Unname et against: convertir en float les lignes digit et remplacer les alpha par des NaN 
unname_column=df.columns[0:11]
against_column=df.columns[12:30]

replace_row_nan(unname_column)
replace_row_nan(against_column)

<ipython-input-8-dde5f1a1fc6a>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[i][j]=np.nan
<ipython-input-8-dde5f1a1fc6a>:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[i][j]=np.nan


Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1  Unnamed: 0.1.1.1  \
0             0             0               0                 0   
1             1             1               1                 1   
2             2             2               2                 2   
3             3             3               3                 3   
4             4             4               4                 4   
..          ...           ...             ...               ...   
796         796           796             796               796   
797         797           797             797               797   
798         798           798             798               798   
799         799           799             799               799   
800         800           800             800               800   

     Unnamed: 0.1.1.1.1 Unnamed: 0.1.1.1.1.1 Unnamed: 0.1.1.1.1.1.1  \
0                     0                    0                    NaN   
1                     1                    1                      1   
2                     2                    2                      2   
3                     3                    3                      3   
4                     4                    4                      4   
..                  ...                  ...                    ...   
796                 796                  796                    796   
797                 797                  797                    797   
798                 798                  798                    798   
799                 799                  799                    799   
800                 800                  800                    NaN   

    Unnamed: 0.1.1.1.1.1.1.1 Unnamed: 0.1.1.1.1.1.1.1.1  \
0                          0                          0   
1                          1                          1   
2                        NaN                          2   
3                          3                          3   
4                          4                          4   
..                       ...                        ...   
796                      796                        796   
797                      797                        797   
798                      798                        798   
799                      799                        799   
800                      800                        800   

    Unnamed: 0.1.1.1.1.1.1.1.1.1  ... percentage_male pokedex_number  \
0                              0  ...            88.1              1   
1                              1  ...            88.1              2   
2                              2  ...            88.1              3   
3                              3  ...            88.1              4   
4                              4  ...            88.1              5   
..                           ...  ...             ...            ...   
796                          796  ...             NaN            797   
797                          797  ...             NaN            798   
798                          798  ...             NaN            799   
799                          799  ...             NaN            800   
800                          800  ...             NaN            801   

    sp_attack sp_defense speed    type1   type2 weight_kg generation  \
0          65         65    45    grass  poison       6.9          1   
1          80         80    60    grass  poison      13.0          1   
2         122        120    80    grass  poison     100.0          1   
3          60         50    65     fire     NaN       8.5          1   
4          80         65    80     fire     NaN      19.0          1   
..        ...        ...   ...      ...     ...       ...        ...   
796       107        101    61    steel  flying     999.9          7   
797        59         31   109    grass   steel       0.1          7   
798        97         53    43     dark  dragon     888.0          7   
799       127         89    79  psychic     NaN     2

In [10]:
df.iloc[:,13:30].head(3)

against_dark against_dragon against_electric against_fairy against_fight  \
0          1.0            NaN              NaN           NaN           0.5   
1          1.0            1.0              NaN           0.5           0.5   
2          1.0            1.0              0.5           0.5           0.5   

  against_fire  against_flying against_ghost  against_grass  against_ground  \
0          2.0             2.0           1.0           0.25             1.0   
1          2.0             2.0           1.0           0.25             1.0   
2          2.0             2.0           1.0           0.25             1.0   

   against_ice  against_normal  against_poison  against_psychic  against_rock  \
0          2.0             1.0             1.0              2.0           1.0   
1          2.0             1.0             1.0              2.0           1.0   
2          2.0             1.0             1.0              2.0           1.0   

   against_steel  against_water  
0            1.0            0.5  
1            1.0            0.5  
2            1.0            0.5

In [11]:
df.capture_rate.unique()

array(['45', '255', '120', '127', '90', '190', '75', '235', '150', '25',
       '170', '50', '200', '100', '180', '60', '225', '30', '35', '3',
       '65', '70', '125', '205', '155', '145', '130', '140', '15', '220',
       '160', '80', '55', '30 (Meteorite)255 (Core)'], dtype=object)

In [12]:
#Traiter la colonne capture_rate:
#Remplacer les lignes qui contiennent la valeur: '30 (Meteorite)255 (Core)' par nan
cap=['capture_rate']
replace_row_nan('capture_rate')
#df[df['capture_rate']=='30 (Meteorite)255 (Core)']['capture_rate']


<ipython-input-8-dde5f1a1fc6a>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[i][j]=np.nan


Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1  Unnamed: 0.1.1.1  \
0             0             0               0                 0   
1             1             1               1                 1   
2             2             2               2                 2   
3             3             3               3                 3   
4             4             4               4                 4   
..          ...           ...             ...               ...   
796         796           796             796               796   
797         797           797             797               797   
798         798           798             798               798   
799         799           799             799               799   
800         800           800             800               800   

     Unnamed: 0.1.1.1.1 Unnamed: 0.1.1.1.1.1 Unnamed: 0.1.1.1.1.1.1  \
0                     0                    0                    NaN   
1                     1                    1                      1   
2                     2                    2                      2   
3                     3                    3                      3   
4                     4                    4                      4   
..                  ...                  ...                    ...   
796                 796                  796                    796   
797                 797                  797                    797   
798                 798                  798                    798   
799                 799                  799                    799   
800                 800                  800                    NaN   

    Unnamed: 0.1.1.1.1.1.1.1 Unnamed: 0.1.1.1.1.1.1.1.1  \
0                          0                          0   
1                          1                          1   
2                        NaN                          2   
3                          3                          3   
4                          4                          4   
..                       ...                        ...   
796                      796                        796   
797                      797                        797   
798                      798                        798   
799                      799                        799   
800                      800                        800   

    Unnamed: 0.1.1.1.1.1.1.1.1.1  ... percentage_male pokedex_number  \
0                              0  ...            88.1              1   
1                              1  ...            88.1              2   
2                              2  ...            88.1              3   
3                              3  ...            88.1              4   
4                              4  ...            88.1              5   
..                           ...  ...             ...            ...   
796                          796  ...             NaN            797   
797                          797  ...             NaN            798   
798                          798  ...             NaN            799   
799                          799  ...             NaN            800   
800                          800  ...             NaN            801   

    sp_attack sp_defense speed    type1   type2 weight_kg generation  \
0          65         65    45    grass  poison       6.9          1   
1          80         80    60    grass  poison      13.0          1   
2         122        120    80    grass  poison     100.0          1   
3          60         50    65     fire     NaN       8.5          1   
4          80         65    80     fire     NaN      19.0          1   
..        ...        ...   ...      ...     ...       ...        ...   
796       107        101    61    steel  flying     999.9          7   
797        59         31   109    grass   steel       0.1          7   
798        97         53    43     dark  dragon     888.0          7   
799       127         89    79  psychic     NaN     2

In [13]:
#Traiter la colonne classification
def rename(row, ch1, ch2):
    for i in df.index:
        row[i]=row[i].replace(ch1,ch2)
    return df
rename(df['classfication'],' Pokémon','')
df['classfication'].head()


<ipython-input-13-0d611f3bb9b6>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  row[i]=row[i].replace(ch1,ch2)


0      Seed
1      Seed
2      Seed
3    Lizard
4     Flame
Name: classfication, dtype: object

In [14]:
#Traiter la colonne name
rename(df['name'],'♂','o')
rename(df['name'],',','')
rename(df['name'],'♀','o')
df['name'].shape


<ipython-input-13-0d611f3bb9b6>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  row[i]=row[i].replace(ch1,ch2)


(801,)

In [15]:
#df=Treatment_1(name_file)
#Traiter la colonne type1 et type2
def create_list_dict(row):
    l=row.unique().tolist()
    l_int= list(range(1,len(l)+1))
    d=dict(zip(l,l_int))
    return l, d

#Declaration de variables
res1 =create_list_dict(df['type1'])
res2 =create_list_dict(df['type2'])
dict_type1=res1[1]
dict_type2=res2[1]

def rename_str_int(row):            
    for i in df.index:
        if row == 'type1':
            for key, val in dict_type1.items():
                if df[row][i] == key:
                    df[row][i]=df[row][i].replace(df[row][i],str(val))
        else:
            for key, val in dict_type2.items():
                if df[row][i] == key:
                    df[row][i]=df[row][i].replace(df[row][i],str(val))
    return df 
rename_str_int('type1')
rename_str_int('type2')



<ipython-input-15-d4db898be098>:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[row][i]=df[row][i].replace(df[row][i],str(val))
<ipython-input-15-d4db898be098>:24: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[row][i]=df[row][i].replace(df[row][i],str(val))


Unnamed: 0  Unnamed: 0.1  Unnamed: 0.1.1  Unnamed: 0.1.1.1  \
0             0             0               0                 0   
1             1             1               1                 1   
2             2             2               2                 2   
3             3             3               3                 3   
4             4             4               4                 4   
..          ...           ...             ...               ...   
796         796           796             796               796   
797         797           797             797               797   
798         798           798             798               798   
799         799           799             799               799   
800         800           800             800               800   

     Unnamed: 0.1.1.1.1 Unnamed: 0.1.1.1.1.1 Unnamed: 0.1.1.1.1.1.1  \
0                     0                    0                    NaN   
1                     1                    1                      1   
2                     2                    2                      2   
3                     3                    3                      3   
4                     4                    4                      4   
..                  ...                  ...                    ...   
796                 796                  796                    796   
797                 797                  797                    797   
798                 798                  798                    798   
799                 799                  799                    799   
800                 800                  800                    NaN   

    Unnamed: 0.1.1.1.1.1.1.1 Unnamed: 0.1.1.1.1.1.1.1.1  \
0                          0                          0   
1                          1                          1   
2                        NaN                          2   
3                          3                          3   
4                          4                          4   
..                       ...                        ...   
796                      796                        796   
797                      797                        797   
798                      798                        798   
799                      799                        799   
800                      800                        800   

    Unnamed: 0.1.1.1.1.1.1.1.1.1  ... percentage_male pokedex_number  \
0                              0  ...            88.1              1   
1                              1  ...            88.1              2   
2                              2  ...            88.1              3   
3                              3  ...            88.1              4   
4                              4  ...            88.1              5   
..                           ...  ...             ...            ...   
796                          796  ...             NaN            797   
797                          797  ...             NaN            798   
798                          798  ...             NaN            799   
799                          799  ...             NaN            800   
800                          800  ...             NaN            801   

    sp_attack sp_defense speed type1 type2 weight_kg generation  is_legendary  
0          65         65    45     1     1       6.9          1             0  
1          80         80    60     1     1      13.0          1             0  
2         122        120    80     1     1     100.0          1             0  
3          60         50    65     2   NaN       8.5          1             0  
4          80         65    80     2   NaN      19.0          1             0  
..        ...        ...   ...   ...   ...       ...        ...           ...  
796       107        101    61    17     3     999.9          7             1  
797        59         31   109     1    12       0.1          7             1  
798        97         53    43    16    16     888.0 

Supprimer les colonnes inutiles 

In [16]:
df=df.drop('abilities', axis=1)
df=df.drop(unname_column, axis=1)
df=df.drop('pokedex_number',axis=1)

In [17]:
df.capture_rate.unique()

array(['45', '255', '120', '127', '90', '190', '75', '235', '150', '25',
       '170', '50', '200', '100', '180', '60', '225', '30', '35', '3',
       '65', '70', '125', '205', '155', '145', '130', '140', '15', '220',
       '160', '80', '55', nan], dtype=object)

Reformater les données de manière pertinente 

In [18]:
def calcul_moy_par_col(row):          
    moy=df[row][df[row].notnull()].astype('float64').mean()
    return moy

In [19]:
def replace_null_with_moy(row):
    if row == 'type1' or row == 'type2':
        df[row] = df[row].replace(np.nan ,0)
        df[row] = df[row].astype(int)
    else:
        df[row] = df[row].replace(np.nan ,calcul_moy_par_col(row))
        df[row] = df[row].astype(float)
    return df

In [20]:
#Remplacer les nulls par la moyenne (ou 0) et convertir le type de la colonne en numéric (float ou int)
for i in against_column:
    replace_null_with_moy(i)

replace_null_with_moy('capture_rate')
replace_null_with_moy('type1')
replace_null_with_moy('type2')
replace_null_with_moy('height_m')
replace_null_with_moy('weight_kg')
replace_null_with_moy('percentage_male')

against_bug  against_dark  against_dragon  against_electric  \
0           1.00      1.000000         0.97971          1.075397   
1           1.00      1.000000         1.00000          1.075397   
2           1.00      1.000000         1.00000          0.500000   
3           0.50      1.000000         1.00000          1.000000   
4           0.50      1.056835         1.00000          1.000000   
..           ...           ...             ...               ...   
796         0.25      1.000000         0.50000          2.000000   
797         1.00      1.000000         0.97971          1.075397   
798         2.00      0.500000         2.00000          0.500000   
799         2.00      2.000000         1.00000          1.000000   
800         0.25      0.500000         0.00000          1.000000   

     against_fairy  against_fight  against_fire  against_flying  \
0         1.051312       0.500000       2.00000        2.000000   
1         0.500000       0.500000       2.00000        2.000000   
2         0.500000       0.500000       2.00000        2.000000   
3         0.500000       1.000000       0.50000        1.000000   
4         0.500000       1.066245       0.50000        1.000000   
..             ...            ...           ...             ...   
796       0.500000       1.000000       1.14747        0.500000   
797       0.500000       2.000000       4.00000        1.000000   
798       1.051312       2.000000       0.50000        1.000000   
799       1.051312       0.500000       1.00000        1.195739   
800       0.500000       1.000000       2.00000        0.500000   

     against_ghost  against_grass  ...        name  percentage_male  \
0         1.000000           0.25  ...   Bulbasaur        88.100000   
1         1.000000           0.25  ...     Ivysaur        88.100000   
2         1.000000           0.25  ...    Venusaur        88.100000   
3         1.000000           0.50  ...  Charmander        88.100000   
4         1.000000           0.50  ...  Charmeleon        88.100000   
..             ...            ...  ...         ...              ...   
796       1.000000           0.25  ...  Celesteela        55.155761   
797       0.982024           0.25  ...     Kartana        55.155761   
798       0.500000           0.50  ...    Guzzlord        55.155761   
799       0.982024           1.00  ...    Necrozma        55.155761   
800       1.000000           0.50  ...    Magearna        55.155761   

     sp_attack  sp_defense  speed  type1  type2  weight_kg  generation  \
0           65          65     45      1      1        6.9           1   
1           80          80     60      1      1       13.0           1   
2          122         120     80      1      1      100.0           1   
3           60          50     65      2      0        8.5           1   
4           80          65     80      2      0       19.0           1   
..         ...         ...    ...    ...    ...        ...         ...   
796        107         101     61     17      3      999.9           7   
797         59          31    109      1     12        0.1           7   
798         97          53     43     16     16      888.0           7   
799        127          89     79     11      0      230.0           7   
800        130         115     65     17      8       80.5           7   

     is_legendary  
0               0  
1               0  
2               0  
3               0  
4               0  
..            ...  
796             1  
797             1  
798             1  
799             1  
800             1  

[801 rows x 39 columns]

In [21]:
df.describe()

against_bug  against_dark  against_dragon  against_electric  \
count   801.000000    801.000000      801.000000        801.000000   
mean      0.997580      1.056835        0.979710          1.075397   
std       0.578855      0.404295        0.330182          0.619600   
min       0.250000      0.250000        0.000000          0.000000   
25%       0.500000      1.000000        1.000000          1.000000   
50%       1.000000      1.000000        1.000000          1.000000   
75%       1.000000      1.056835        1.000000          1.075397   
max       4.000000      4.000000        2.000000          4.000000   

       against_fairy  against_fight  against_fire  against_flying  \
count     801.000000     801.000000    801.000000      801.000000   
mean        1.051312       1.066245      1.147470        1.195739   
std         0.450321       0.696435      0.677869        0.592243   
min         0.250000       0.000000      0.250000        0.250000   
25%         1.000000       0.500000      0.500000        1.000000   
50%         1.000000       1.000000      1.000000        1.000000   
75%         1.051312       1.066245      1.147470        1.195739   
max         4.000000       4.000000      4.000000        4.000000   

       against_ghost  against_grass  ...          hp  percentage_male  \
count     801.000000     801.000000  ...  801.000000       801.000000   
mean        0.982024       1.034020  ...   68.958801        55.155761   
std         0.535032       0.788896  ...   26.576015        18.980069   
min         0.000000       0.250000  ...    1.000000         0.000000   
25%         1.000000       0.500000  ...   50.000000        50.000000   
50%         1.000000       1.000000  ...   65.000000        50.000000   
75%         1.000000       1.000000  ...   80.000000        55.155761   
max         4.000000       4.000000  ...  255.000000       100.000000   

        sp_attack  sp_defense       speed       type1       type2   weight_kg  \
count  801.000000  801.000000  801.000000  801.000000  801.000000  801.000000   
mean    71.305868   70.911361   66.334582    7.024969    4.141074   61.378105   
std     32.353826   27.942501   28.907662    4.782716    5.338663  107.979179   
min     10.000000   20.000000    5.000000    1.000000    0.000000    0.100000   
25%     45.000000   50.000000   45.000000    3.000000    0.000000    9.300000   
50%     65.000000   66.000000   65.000000    5.000000    1.000000   28.500000   
75%     91.000000   90.000000   85.000000   11.000000    8.000000   61.500000   
max    194.000000  230.000000  180.000000   18.000000   19.000000  999.900000   

       generation  is_legendary  
count  801.000000    801.000000  
mean     3.690387      0.087391  
std      1.930420      0.282583  
min      1.000000      0.000000  
25%      2.000000      0.000000  
50%      4.000000      0.000000  
75%      5.000000      0.000000  
max      7.000000      1.000000  

[8 rows x 36 columns]

Packager votre traitement sous forme de fonctions réutilisables pour un autre set de données de même structure 
(généralisation maximale)

In [22]:
#df.to_csv('test.csv', index = False)